### ISSUE TITLE
v2.0.0 - Typings issue for Breakpoint and useCurrentWidth/useCurrentBreakpointName

### ISSUE THREAD

**Issue by dbryantm** [flexdinesh/react-socks#23]
The new TypeScript typings break custom breakpoint names for Breakpoint (i.e. sm, md, lg, etc). I currently get the following error:

```
Type '{ children: Element; sm: true; down: true; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Breakpoint> & Readonly<Props> & Readonly<{ children?: ReactNode; }>'.
  Property 'sm' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Breakpoint> & Readonly<Props> & Readonly<{ children?: ReactNode; }>'.
```

Just a suggestion, but either have a breakpoint prop that accepts a string with the following typing (also probably more maintainable):

```
interface Props {
  children?: React.ReactNode;
  breakpoint?: string;
  up?: boolean;
  down?: boolean;
  only?: boolean;
  tagName?: string;
  className?: string;
}
```

Or change the Props typing to the following (which unfortunately lets more in but doesn't restrict the use of the library):

```
interface Props<T extends any> {
  children?: React.ReactNode;
  up?: boolean;
  down?: boolean;
  only?: boolean;
  tagName?: string;
  className?: string;
}
```

Also, you're missing typings for useCurrentWidth and useCurrentBreakpointName:

```
export function useCurrentWidth(): number;
export function userCurrentBreakpointName(): string;
```

**Comment by guschnwg** [comment]
My breakpoints are `small`, `medium` and `large`, somewhere in my app I added the file `socks.d.ts` with the following code:

```
namespace ReactSocks {
  interface Props {
    children?: React.ReactNode;
    up?: boolean;
    down?: boolean;
    only?: boolean;
    tagName?: string;
    className?: string;
    small?: boolean;
    medium?: boolean;
    large?: boolean;
  }
}
```

**Comment by dbryantm** [comment]
I've already done something very similar. Though it would probably be good to either document it on the README or update the typings to make them work as advertised. I was able to resolve it quickly though not everyone may have the same experience.

**Comment by guschnwg** [comment]
Totally agreed, as my suggestion was only a temporary solution.

As my react app was not compiling with that error you mentioned, that temporary solution solved the issue for me, and could solve (tomporary) to another person who read your issue and was hurting only for the first part

**Comment by flexdinesh** [comment]
@dbryantm Thanks for the report. Would you be interested in improving this? Quite honestly TypeScript is a bit out of my depth.

**Comment by flexdinesh** [comment]
@nutboltu Perhaps you'd like to add something to this discussion?

**Comment by nutboltu** [comment]
@dbryantm and @flexdinesh  It seems, we are expecting any props ([see here](https://github.com/flexdinesh/react-socks/blob/e49a6556969c85c3223ff803593770cf467d3864/src/Breakpoint/Breakpoint.js#L25)) and replace it with breakpoint. 
In this case consumer can pass any props (`sm`, `md`, `lg`, `small`, `medium`, `large` ).

**Possible Solution: 1**

As @dbryantm mentioned, we can introduce a new props `breakpoint` where consumer can pass custom breakpoint.
```
<Breakpoint breakpoint="small" down>
</Breakpoint>
```
**Possible Solution: 2**
We can add an unknown property in the Props

```
interface Props {
      children?: React.ReactNode;
      up?: boolean;
      down?: boolean;
      only?: boolean;
      tagName?: string;
      className?: string;
      [breakpoint: string]: string
    }
```
Let me know your thoughts.

**Comment by flexdinesh** [comment]
@nutboltu When we started, the biggest highlight of this lib was its sugary syntax (being able to pass multiple breakpoints with shorthands) and I'd like to keep it that way unless there's a performance or demonstrated bad practice in play.

Solution 2 seems convincing to me. It will keep TS consumers happy while retaining the syntax. What do you all think? @dbryantm @guschnwg @nutboltu

**Comment by dbryantm** [comment]
@nutboltu It's actually a relatively common practice in library distribution I believe unless your intent is to be opinionated with how users interact with your software. My radio silence was due to me trying to think of better options. I still feel something like the following will keep the nice syntax while not opening it up too much:

```
interface Props<T = {[key: string]: boolean}> {
  children?: React.ReactNode;
  up?: boolean;
  down?: boolean;
  only?: boolean;
  tagName?: string;
  className?: string;
}
```

Let me know your thoughts. I could submit a PR for this and the two hooks that need to export their typings also.

**Comment by nutboltu** [comment]
+1  for Solution 2 then 🎉 .

**Comment by flexdinesh** [comment]
@dbryantm Sounds great. Please go ahead with PR. 🎉

**Comment by dbryantm** [comment]
Just an FYI, if you look at the following example:

[TypeScript Playground - Proposed Solution Issues](https://www.typescriptlang.org/play/index.html#code/PTAEEsHMDsHsCcCmoAuALZAlRBDAxigHTb4oBysAJsovPAqDgM4QoDkLAsAFAihwoIAWwAOCFDmiCAZg3TgW1IbGhMU8HCnArCPHnxH0xTRJVBNYAGwCuWlT3BTa0-MgAKR2CcofYIlgDePKCgeGjglpRI0AD8AFygJATEuAQU1ADcwaDWIvGgAEawVrjQWdwhlLAA7rEJRSWS5SEqlgCe+Q2Wpc2oOJBkOEKI+WrwjpC9eJbMTIPDo+oTvQDaBUg4ANZijigJYxMAuvXF3U08AL56vGDVCJsT5la22tAOTvAueMgA6vcTvn8AB4ACqgRAADxQiGglECoBWm0QbX2S2gkGOjGgbVAFwAfKAghVQuFItF8kkiJT0oherlOqcetkqrUGY0ytlWh0TuzehIBkMRqjxuipjMmHNBYsRZNLkA)

I think we're going to have to go with something closer to my first suggested solution. I'll still go ahead and submit a PR but would like to get your thoughts using that as evidence for my findings. Thanks in advance!

**Comment by flexdinesh** [comment]
@nutboltu Would you be able to review the PR?

**Comment by nutboltu** [comment]
@dbryantm Thanks for fixing the unknown props issue in a cleaner way. it looks good to me.

**Comment by flexdinesh** [comment]
Changes published in `v2.0.1`. Thanks @dbryantm @nutboltu @guschnwg! 🎉
